src/index.js

```diff
@@ -61,83 +56,61 @@ class IpfsRepo {
   }
 
   /**
    * Open the repo. If the repo is already open no action will be taken.
    * If the repo is not initialized it will return an error.
    *
-   * @param {function(Error)} callback
-   * @returns {void}
+   * @returns {Promise<void>}
    */
-  open (callback) {
+  async open () {
     if (!this.closed) {
-      setImmediate(() => callback(new Error('repo is already open')))
-      return // early
+      throw new Error('repo is already open')
     }
     log('opening at: %s', this.path)
 
     // check if the repo is already initialized
-    waterfall([
-      (cb) => this.root.open(ignoringAlreadyOpened(cb)),
-      (cb) => this._isInitialized(cb),
-      (cb) => this._openLock(this.path, cb),
-      (lck, cb) => {
-        log('aquired repo.lock')
-        this.lockfile = lck
-        cb()
-      },
-      (cb) => {
-        log('creating datastore')
-        this.datastore = backends.create('datastore', path.join(this.path, 'datastore'), this.options)
-        log('creating blocks')
-        const blocksBaseStore = backends.create('blocks', path.join(this.path, 'blocks'), this.options)
-        blockstore(
-          blocksBaseStore,
-          this.options.storageBackendOptions.blocks,
-          cb)
-      },
-      (blocks, cb) => {
-        this.blocks = blocks
-        cb()
-      },
-      (cb) => {
-        log('creating keystore')
-        this.keys = backends.create('keys', path.join(this.path, 'keys'), this.options)
-        cb()
-      },
-
-      (cb) => {
-        this.closed = false
-        log('all opened')
-        cb()
+    try {
+      await this.root.open()
+      const initialized = await this._isInitialized()
+      if (!initialized) {
+        throw Object.assign(new Error('repo is not initialized yet'),
+          {
+            code: ERRORS.ERR_REPO_NOT_INITIALIZED,
+            path: this.path
+          })
       }
-    ], (err) => {
+      this.lockfile = await this._openLock(this.path)
+      log('aquired repo.lock')
+      log('creating datastore')
+      this.datastore = backends.create('datastore', path.join(this.path, 'datastore'), this.options)
+      log('creating blocks')
+      const blocksBaseStore = backends.create('blocks', path.join(this.path, 'blocks'), this.options)
+      this.blocks = await blockstore(blocksBaseStore, this.options.storageBackendOptions.blocks)
+      log('creating keystore')
+      this.keys = backends.create('keys', path.join(this.path, 'keys'), this.options)
+      this.closed = false
+      log('all opened')
+    } catch (err) {
       if (err && this.lockfile) {
-        this._closeLock((err2) => {
-          if (!err2) {
-            this.lockfile = null
-          } else {
-            log('error removing lock', err2)
-          }
-          callback(err)
-        })
-      } else {
-        callback(err)
+        try {
+          this._closeLock()
+        } catch (err2) {
+          log('error removing lock', err2)
+        }
+        this.lockfile = null
```

Comment:
I think you only want to set `this.lockfile` to null if there was not an error thrown from `this._closeLock()`